subscriber/common/sink/sink.go

```diff
@@ -0,0 +1,138 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sink
+
+import (
+	"fmt"
+	"github.com/uber/aresdb/client"
+	"github.com/uber/aresdb/memstore"
+	memCom "github.com/uber/aresdb/memstore/common"
+	"github.com/uber/aresdb/subscriber/common/rules"
+	"github.com/uber/aresdb/utils"
+	"strings"
+	"unsafe"
+)
+
+// Sink is abstraction for interactions with downstream storage layer
+type Sink interface {
+	// Cluster returns the DB cluster name
+	Cluster() string
+
+	// Save will save the rows into underlying database
+	Save(destination Destination, rows []client.Row) error
+
+	// Shutdown will close the connections to the database
+	Shutdown()
+}
+
+// Destination contains the table and columns that each job is storing data into
+// also records the behavior when encountering key errors
+type Destination struct {
+	// Table is table name
+	Table string
+	// ColumnNames are the list of column names after sorted
+	ColumnNames []string
+	// PrimaryKeys maps primary key columnName to its columnID after sorted
+	PrimaryKeys map[string]int
+	// PrimaryKeysInSchema maps primary key columnName to its columnID defined in schema
+	PrimaryKeysInSchema map[string]int
+	// AresUpdateModes defines update modes
+	AresUpdateModes []memCom.ColumnUpdateMode
+	// NumShards is the number of shards in the aresDB cluster
+	NumShards uint32
+}
+
+func Shard(rows []client.Row, destination Destination, jobConfig *rules.JobConfig) map[uint32][]client.Row {
+	if destination.NumShards == 0 || destination.NumShards == 1 {
+		// in this case, there is no sharding in this aresDB cluster
+		return nil
+	}
+
+	shards := make(map[uint32][]client.Row)
+	for i := uint32(0); i < destination.NumShards; i++ {
+		shards[i] = make([]client.Row, 0, len(rows))
+	}
+
+	for _, row := range rows {
+		// convert primaryKey to byte array
+		pk := make([]byte, jobConfig.GetPrimaryKeyBytes())
+		err := GetPrimaryKeyBytes(row, destination, jobConfig, &pk)
+		if err != nil {
+			utils.StackError(err, "Failed to convert primaryKey to byte array for row: %v", row)
+			continue
+		}
+
+		// calculate shard
+		shardID := shardFn(pk, destination.NumShards)
+		shards[shardID] = append(shards[shardID], row)
+	}
+	return shards
+}
+
+func shardFn(key []byte, numShards uint32) uint32 {
+	return utils.Murmur3Sum32(unsafe.Pointer(&key[0]), len(key), 0) % numShards
+}
+
+func GetPrimaryKeyBytes(row client.Row, destination Destination, jobConfig *rules.JobConfig, key *[]byte) error {
+	primaryKeyValues := make([]memCom.DataValue, len(destination.PrimaryKeys))
+	var err error
+	var strBytes []byte
+	i := 0
+	for columnName, columnID := range destination.PrimaryKeys {
+		columnIDInSchema := destination.PrimaryKeysInSchema[columnName]
+		if jobConfig.AresTableConfig.Table.Columns[columnIDInSchema].Type == memCom.DataTypeName[memCom.SmallEnum] ||
+			jobConfig.AresTableConfig.Table.Columns[columnIDInSchema].Type == memCom.DataTypeName[memCom.BigEnum] {
```

Comment:
under Column there is a call IsEnum() can be used to determine enum column

Reply:
great. will replace